pkg/dns/dns.go

```diff
@@ -0,0 +1,452 @@
+package dns
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/containous/maesh/pkg/k8s"
+	"github.com/containous/traefik/v2/pkg/safe"
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	kubeerror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Provider represents a DNS provider.
+type Provider int
+
+// Supported DNS providers.
+const (
+	UnknownDNS Provider = iota
+	CoreDNS
+	KubeDNS
+)
+
+var (
+	supportedCoreDNSVersions = []string{
+		"1.3",
+		"1.4",
+		"1.5",
+		"1.6",
+	}
+	coreFileHeader  = "#### Begin Maesh Block"
+	coreFileTrailer = "#### End Maesh Block"
+)
+
+// Client holds the client for interacting with the k8s DNS system.
+type Client struct {
+	client k8s.Client
+	log    logrus.FieldLogger
+}
+
+// NewClient returns an initialized DNSClient object.
+func NewClient(log logrus.FieldLogger, client k8s.Client) *Client {
+	return &Client{
+		client: client,
+		log:    log,
+	}
+}
+
+// CheckDNSProvider checks that the DNS provider that is deployed in the cluster
+// is supported and returns it.
+func (c *Client) CheckDNSProvider() (Provider, error) {
+	c.log.Info("Checking DNS provider")
+
+	match, err := c.coreDNSMatch()
+	if err != nil {
+		return UnknownDNS, err
+	}
+
+	if match {
+		return CoreDNS, nil
+	}
+
+	match, err = c.kubeDNSMatch()
+	if err != nil {
+		return UnknownDNS, err
+	}
+
+	if match {
+		return KubeDNS, nil
+	}
+
+	return UnknownDNS, fmt.Errorf("no core DNS service available for installing maesh: %w", err)
+}
+
+func (c *Client) coreDNSMatch() (bool, error) {
+	c.log.Info("Checking CoreDNS")
+
+	deployment, err := c.client.KubernetesClient().AppsV1().Deployments(metav1.NamespaceSystem).Get("coredns", metav1.GetOptions{})
+	if kubeerror.IsNotFound(err) {
+		c.log.Debugf("CoreDNS does not exist in namespace %q", metav1.NamespaceSystem)
+		return false, nil
+	}
+
+	if err != nil {
+		return false, fmt.Errorf("unable to get deployment %q in namesapce %q: %w", "coredns", metav1.NamespaceSystem, err)
+	}
+
+	var version string
+
+	for _, container := range deployment.Spec.Template.Spec.Containers {
+		if container.Name != "coredns" {
+			continue
+		}
+
+		sp := strings.Split(container.Image, ":")
+		version = sp[len(sp)-1]
+	}
+
+	if !isCoreDNSVersionSupported(version) {
+		return false, fmt.Errorf("unsupported CoreDNS version %q, (supported versions are: %s)", version, strings.Join(supportedCoreDNSVersions, ","))
+	}
+
+	c.log.Info("CoreDNS match")
+
+	return true, nil
+}
+
+func isCoreDNSVersionSupported(versionLine string) bool {
+	for _, v := range supportedCoreDNSVersions {
+		if strings.Contains(versionLine, v) || strings.Contains(versionLine, "v"+v) {
```

Comment:
This condition is not needed?
In both cases the string contains `v`.

```suggestion
		if strings.Contains(versionLine, v) {
```